ref-exchange/src/multi_fungible_token.rs

```diff
@@ -0,0 +1,184 @@
+                    .deposited_amounts
+                    .get(receiver_id)
+                    .expect(ERR10_ACC_NOT_REGISTERED);
+                sender_account.sub(&token_id, amount);
+                receiver_account.add(&token_id, amount);
+                self.deposited_amounts.insert(&sender_id, &sender_account);
+                self.deposited_amounts
+                    .insert(&receiver_id, &receiver_account);
+                log!(
+                    "Transfer {}: {} from {} to {}",
+                    token_id,
+                    amount,
+                    sender_id,
+                    receiver_id
+                );
+            }
+        }
+        if let Some(memo) = memo {
+            log!("Memo: {}", memo);
+        }
+    }
+
+    /// Transfer one of internal tokens: LP or balances.
+    /// `token_id` can either by account of the token or pool number.
+    #[payable]
+    pub fn mft_transfer(
+        &mut self,
+        token_id: String,
+        receiver_id: ValidAccountId,
+        amount: U128,
+        memo: Option<String>,
+    ) {
+        assert_one_yocto();
+        self.internal_transfer(
+            token_id,
+            &env::predecessor_account_id(),
+            receiver_id.as_ref(),
+            amount.0,
+            memo,
+        );
+    }
+
+    #[payable]
+    pub fn mft_transfer_call(
+        &mut self,
+        token_id: String,
+        receiver_id: ValidAccountId,
+        amount: U128,
+        memo: Option<String>,
+        msg: String,
+    ) -> PromiseOrValue<U128> {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        self.internal_transfer(
+            token_id.clone(),
+            &sender_id,
+            receiver_id.as_ref(),
+            amount.0,
+            memo,
+        );
+        ext_share_token_receiver::mft_on_transfer(
+            token_id.clone(),
+            sender_id.clone(),
+            amount,
+            msg,
+            receiver_id.as_ref(),
+            NO_DEPOSIT,
+            env::prepaid_gas() - GAS_FOR_FT_TRANSFER_CALL,
+        )
+        .then(ext_self::mft_resolve_transfer(
+            token_id,
+            sender_id,
+            receiver_id.into(),
+            amount,
+            &env::current_account_id(),
+            NO_DEPOSIT,
+            GAS_FOR_RESOLVE_TRANSFER,
+        ))
+        .into()
+    }
+
+    #[private]
+    pub fn mft_resolve_transfer(
+        &mut self,
+        token_id: String,
+        sender_id: AccountId,
+        receiver_id: &AccountId,
+        amount: U128,
+    ) -> U128 {
+        let unused_amount = match env::promise_result(0) {
+            PromiseResult::NotReady => unreachable!(),
+            PromiseResult::Successful(value) => {
+                if let Ok(unused_amount) = near_sdk::serde_json::from_slice::<U128>(&value) {
+                    std::cmp::min(amount.0, unused_amount.0)
+                } else {
+                    amount.0
```

Comment:
`amount.0` should be `0`

Reply:
This is failure to parse the returning value, which is considered a failure.